File: Labs/Lab 7/Main.java
public int findTheCity(int n, int[][] edges, int distanceThreshold) {
    
    ArrayList<int[]>[] graph = new ArrayList[n];
  
    for(int i = 0; i < n; i++){
        graph[i] = new ArrayList<>();
    }
  
    for(int[] e: edges){
        graph[e[0]].add(new int[] {e[1],e[2]});
        graph[e[1]].add(new int[] {e[0],e[2]}); 
    }
  
    int city = 0;
    int min = n+1;
    
    for(int i=0;i<n;i++){
        int count = bfs(graph, i, distanceThreshold);
        if(count <= min){
            city = i;
            min = count;
        }
    }
  
    return city;
}

private int bfs(ArrayList<int[]>[] graph, int s, int k){
    
    PriorityQueue<int[]> q = new PriorityQueue<>((a,b)->a[1]-b[1]);
    q.add(new int[] {s,0});
    boolean[] visited = new boolean[graph.length];
    int nodeCount = -1; 
  
    while(q.size() != 0){
        int[] temp = q.poll();
      
        if (visited[temp[0]]) {
          continue;
        }
      
        visited[temp[0]] = true;
        nodeCount++;
        
        for(int[] x: graph[temp[0]]){
            if(visited[x[0]] != true && x[1] + temp[1] <= k) { 
                q.add(new int[] {x[0], x[1] + temp[1]});
            }
        }
    }
  
    return nodeCount; 
}

File: Labs/Lab 7/Main.py
def findTheCity(self, n: int, edges: List[List[int]], distanceThreshold: int) -> int:
    def bfs(graph, s, k):
        q = [(0, s)]  # (distance, node)
        visited = [False] * len(graph)
        node_count = -1
        
        while q:
            distance, node = heapq.heappop(q)
            
            if visited[node]:
                continue
            
            visited[node] = True
            node_count += 1
            
            for neighbor, weight in graph[node]:
                if not visited[neighbor] and weight + distance <= k:
                    heapq.heappush(q, (weight + distance, neighbor))
        
        return node_count
    
    graph = [[] for _ in range(n)]
    
    for e in edges:
        graph[e[0]].append((e[1], e[2]))
        graph[e[1]].append((e[0], e[2]))
    
    city = 0
    min_count = n + 1
    
    for i in range(n):
        count = bfs(graph, i, distanceThreshold)
        if count <= min_count:
            city = i
            min_count = count
    
    return city
